nucypher/blockchain/eth/sol/source/contracts/WorkLock.sol

```diff
@@ -0,0 +1,147 @@
+        uint256 _refundRate,
+        uint16 _lockedPeriods
+    )
+        public
+    {
+        require(_token.totalSupply() > 0 &&
+            _escrow.secondsPerPeriod() > 0 &&
+            _endBidDate > _startBidDate &&
+            _endBidDate > block.timestamp &&
+            _depositRate > 0 &&
+            _refundRate > 0 &&
+            _lockedPeriods >= _escrow.minLockedPeriods());
+        token = _token;
+        escrow = _escrow;
+        startBidDate = _startBidDate;
+        endBidDate = _endBidDate;
+        minAllowableLockedTokens = _escrow.minAllowableLockedTokens();
+        maxAllowableLockedTokens = _escrow.maxAllowableLockedTokens();
+        depositRate = _depositRate;
+        refundRate = _refundRate;
+        lockedPeriods = _lockedPeriods;
+    }
+
+    /**
+    * @notice Bid for tokens by transferring ETH
+    **/
+    function bid() public payable returns (uint256 newClaimedTokens) {
+        require(block.timestamp >= startBidDate && block.timestamp <= endBidDate,
+            "Bid is open during a certain period");
+        WorkInfo storage info = workInfo[msg.sender];
+        info.depositedETH = info.depositedETH.add(msg.value);
+        uint256 claimedTokens = info.depositedETH.mul(depositRate);
+        require(claimedTokens >= minAllowableLockedTokens && claimedTokens <= maxAllowableLockedTokens,
+            "Claimed tokens must be within the allowed limits");
+        newClaimedTokens = msg.value.mul(depositRate);
+        allClaimedTokens = allClaimedTokens.add(newClaimedTokens);
+        require(allClaimedTokens <= token.balanceOf(address(this)),
+            "Not enough tokens in the contract");
+        emit Bid(msg.sender, msg.value, newClaimedTokens);
+    }
+
+    /**
+    * @notice Claimed tokens will be deposited and locked as stake in the MinersEscrow contract.
+    **/
+    function claim() public returns (uint256 claimedTokens) {
+        require(block.timestamp >= endBidDate, "Claiming tokens allowed after bidding is over");
+        WorkInfo storage info = workInfo[msg.sender];
+        require(!info.claimed, "Tokens are already claimed");
+        info.claimed = true;
+        claimedTokens = info.depositedETH.mul(depositRate);
+        info.workDone = escrow.setWorkMeasurement(msg.sender, true);
+        token.approve(address(escrow), claimedTokens);
+        escrow.deposit(msg.sender, claimedTokens, lockedPeriods);
+        emit Claimed(msg.sender, claimedTokens);
+    }
+
+    /**
+    * @notice Refund ETH for the work done
+    **/
+    function refund() public returns (uint256 refundETH) {
+        WorkInfo storage info = workInfo[msg.sender];
+        require(info.claimed, "Tokens are not claimed");
+        require(info.depositedETH > 0, "Nothing deposited");
+        uint256 currentWork = escrow.getWorkDone(msg.sender);
+        uint256 workDone = currentWork.sub(info.workDone);
+        require(workDone > 0, "No work has been done.");
+        refundETH = workDone.div(refundRate);
```

Comment:
Currently inflationary rewards earned (in NU) since the last ETH withdrawal are the primary input to generate `workDone`. Then, unless I'm mistaken, the initial sum of NU acquired via the WorkLock (`claimedTokens`) is not incorporated into the calculation to produce `refundETH `.

Given that it is measured as rewards earned in absolute terms (i.e. # of NU tokens), the ability/power to produce 'work' is varies enormously between large and small stakers. Hence a constant (`refundRate`) cannot suitably cover the range of different stake sizes – large stakers will be able to get their ETH back many times faster than small stakers.

Reply:
I'm not sure that I understand whole idea, few moments:
* `refundETH` and `claimedTokens` are already connected through `refundRate`. If `refundRate` == `depositRate` then reward tokens have same value as claimed tokens.
* `relativeWork` in suggetions almost allways zero, because `workDone` will be less than claimed token initially
* Speed of refund in percents will be the same independatly of initial stake, so depending on state of whole net big and small stakers need same amount of time to get 100% of ETH back

Reply:
@arjunhassard percent-wise, there is no difference between large and small stakers, really.

The only problem I currently see is the minimal stake (e.g. very small stakers either cannot stake, or we have gas wars)